### PR TITLE
feat(cli): enable the new progress tracker for deployments

### PR DESCRIPTION
**Pull request by efekarakus**
_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: internal/pkg/deploy/cloudformation/cloudformation.go [hunk 1]
```diff
@@ -190,6 +238,12 @@ func (cf CloudFormation) changeRenderers(in changeRenderersInput) ([]progress.Re
 		}
 		var renderer progress.Renderer
 		switch {
+		case aws.StringValue(change.ResourceChange.ResourceType) == ecsServiceResourceType:
+			renderer = progress.ListeningECSServiceResourceRenderer(in.stackStreamer, cf.ecsClient, logicalID, description, progress.ECSServiceRendererOpts{
```
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/cloudformation.go:242]: Slick!

File: internal/pkg/cli/job_deploy.go [hunk 2]
```diff
@@ -271,17 +272,10 @@ func (o *deployJobOpts) deployJob(addonsURL string) error {
 	if err != nil {
 		return err
 	}
-	o.spinner.Start(
```
Comment (iamhopaul123) [inline internal/pkg/cli/job_deploy.go:274]: No more spinner 💯

File: internal/pkg/stream/ecs.go [hunk 3]
```diff
@@ -43,7 +44,14 @@ func (d ECSDeployment) isPrimary() bool {
 }
 
 func (d ECSDeployment) done() bool {
-	return d.RolloutState != rollOutInProgress
+	switch d.RolloutState {
+	case rollOutFailed:
+		return true
+	case rollOutCompleted:
+		return d.DesiredCount == d.RunningCount
```
Comment (iamhopaul123) [inline internal/pkg/stream/ecs.go:51]: Is it possible that the deployment is `COMPLETED` but the tasks are not all running?

File: internal/pkg/term/progress/cloudformation.go [hunk 4]
```diff
@@ -149,22 +149,24 @@ type ecsServiceResourceComponent struct {
...
 	for ev := range c.cfnStream {
 		if c.logicalID != ev.LogicalResourceID {
 			continue
 		}
-		curStatus := ev.ResourceStatus
-		if curStatus != prevStatus && cloudformation.StackStatus(curStatus).UpsertInProgress() {
+		if cloudformation.StackStatus(ev.ResourceStatus).UpsertInProgress() {
+			if ev.PhysicalResourceID == "" {
+				// New service creates receive two "CREATE_IN_PROGRESS" events.
+				// The first event doesn't have a service name yet, the second one has.
```
Comment (iamhopaul123) [inline internal/pkg/term/progress/cloudformation.go:160]: Hmm is this behavior documented anywhere?

**Review by iamhopaul123 — COMMENTED** [review]
Super cool feature 🔥 Can't wait to use it for my copilot deployment.

**Review by iamhopaul123 — APPROVED** [review]

File: internal/pkg/stream/ecs.go (hunk 3 above)
Comment (efekarakus) [reply internal/pkg/stream/ecs.go:51]: Yeah T_T when you change the manifest for example and set `count: 3` from `count: 2`. Then the rollout is completed  from the beginning, but the `desiredCount=3`, `pendingCount=1` and `runningCount=2`

File: internal/pkg/term/progress/cloudformation.go (hunk 4 above)
Comment (efekarakus) [reply internal/pkg/term/progress/cloudformation.go:160]: I couldn't find anything in official docs 🤷  but from observation, it seems we first go into `CREATE_IN_PROGRESS` when there is no service yet, then a second event occurs which actually calls [`CreateService`](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_CreateService.html) and returns to us the service ARN:
Screen Shot 2021-01-27 at 5 20 38 PM

Comment (iamhopaul123) [reply internal/pkg/term/progress/cloudformation.go:160]: yeah seems like it applies to the other resources creation as well.

**Review by huanjani — APPROVED** [review]
!WoW!
!YaY!

**Pull request merged**